Feat/image 이미지 CRUD 및 테스트코드 구현

minjonyyy opened the pull request:
## 🔗 Issue Number
close #8


## 📝 작업 내역
<!--- 구현 내용 및 변경 사항, 관련 이슈에 대해 간단하게 작성해주세요.-->
- [X] 이미지 업로드 기능 
: targetType과 targetId로 구분된 폴더에 s3에 업로드 후 url 반환
- [X] 이미지 수정 기능 
: 기존 폴더에 있던 객체 삭제 -> 새로 업로드 후 url 반환
- [X] 이미지 삭제 기능 
: 해당 폴더에 있는 이미지 전체 삭제


## 💡 PR 특이사항
<!--- PR을 볼 때 팀원에게 알려야 할 특이사항, 논의해야할 부분 등을 알려주세요.-->
* 어제(7일) 피드백 반영 완료
* 포스트맨으로 테스트 완료
* 테스트코드 정상 작동 확인 완료

원래 커밋을 더 나눴어야 하는데...  조금 몰아져 있습니다
커밋에 이슈번호도 몇번 까먹고 안 붙였습니다.... (다음부턴 잘 붙이겠습니다!!!!!)

minjonyyy committed:
chore : aws 의존성 추가

minjonyyy committed:
feat : aws s3 config 파일 추가

minjonyyy committed:
feat : 이미지 업로드, 업데이트, 삭제 기능 구현 #8

minjonyyy committed:
feat : 이미지 CRUD 에러코드 추가 #8

minjonyyy committed:
feat : 특정 확장자만 업로드 가능하게 추가

minjonyyy committed:
feat : dto에 @Schema 어노테이션 추가, @Builder 추가 #8

minjonyyy committed:
feat : 에러코드 추가 #8

minjonyyy committed:
feat : dto에 @Schema 어노테이션 추가 #8

minjonyyy committed:
refactor : url 매핑 수정

minjonyyy committed:
style : 사용하지 않는 import 삭제 #8

minjonyyy committed:
refactor : 서비스로직 구현체로 분리 / feat : 파일 확장자 검증 추가 #8

minjonyyy committed:
test : 서비스 로직 테스트코드 작성 #8

minjonyyy committed:
chore : application-aws.yml 내용 추가 #8

juno-soodal commented on src/main/java/com/parkez/common/config/S3Config.java:
```diff
@@ -0,0 +1,29 @@
+    @Value("${cloud.aws.credentials.secret-key}")
+    private String secretKey;
+    @Value("${cloud.aws.region.static}")
+    private String region;
```
properties가 3개 이상이면 별도의 ConfigurationProperties로 관리하면 좋습니다.

juno-soodal commented on src/main/java/com/parkez/common/image/ImageController.java:
```diff
@@ -0,0 +1,47 @@
+public class ImageController {
+
+    private final S3ImageService imageService;
+
```
지금은 S3를 사용해서 이미지를 관리하지만 향후 바뀔 가능성도 있기 때문에 인터페이스를 의존해서 만들어주면 좋습니다.

juno-soodal commented on src/main/java/com/parkez/common/image/ImageController.java:
```diff
@@ -0,0 +1,47 @@
+    private final S3ImageService imageService;
+
+    @PostMapping
+    public Response<List<String>> upload(
```
응답 결과값은 List<String>이 맞지만 추후에 응답 결과값을 확장할 수도 있기 때문에
별도의 응답 dto를 만들어서 감싸주면 좋습니다.

juno-soodal commented on src/main/java/com/parkez/common/image/ImageController.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    @PutMapping
+    public Response<List<String>> update(
```
응답 결과값은 List<String>이 맞지만 추후에 응답 결과값을 확장할 수도 있기 때문에
별도의 응답 dto를 만들어서 감싸주면 좋습니다.

juno-soodal commented on src/main/java/com/parkez/common/image/S3ImageService.java:
```diff
@@ -0,0 +1,153 @@
+
+        List<String> imageUrls = new ArrayList<>();
+        ImageTargetType targetType = ImageTargetType.fromString(request.getTargetType());
+
```
String 타입을 enum으로 바꾸는 메서드인거 같은데
fromString 보다는 of로 바꾸는게 좋아 보입니다.
뭔가 내부 구현을 너무 드러내는 이름 같습니다.

juno-soodal commented on src/main/java/com/parkez/common/image/S3ImageService.java:
```diff
@@ -0,0 +1,153 @@
+        return AllowedExtension.isAllowedExtension(extension);
+    }
+
+    private String getFileExtension(MultipartFile file){
+        String originalFilename = file.getOriginalFilename();
+        if(originalFilename!=null){
+
+            int index = originalFilename.lastIndexOf('.');
+            if (index == -1 || index == originalFilename.length() - 1) {
+                return ""; // 확장자가 없거나, .으로 끝나는 경우
+            }
+            return originalFilename.substring(index + 1).toLowerCase();
+
+        } throw new ParkingEasyException(ImageErrorCode.IMAGE_IS_NULL);
```
확장자를 추출하는 메서드로 보이는데
get보다는 extract로 바꾸는게 추출하는 의미가 나타나서 좋아 보입니다.

확장자를 추출할때는 file 보다는
string filename을 파라미터로 받는게 좋아 보입니다.
추출 자체는 string 타입만 필요하니까요

얼리 리턴형태로 코드를 수정하는게 좀 더 가독성 측면에서 좋아 보입니다.
예외를 먼저 처리하고
최종 리턴은 성공형태로 수정하면 좋아보입니다.

juno-soodal commented on src/main/java/com/parkez/common/image/dto/ImageDeleteRequest.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.List;
+
+@Getter
+@NoArgsConstructor
+public class ImageDeleteRequest {
+
```
request  자체에 대한 Schema도 작성하면 좋습니다.

juno-soodal commented on src/main/java/com/parkez/common/image/dto/ImageUploadRequest.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.List;
+
+@Getter
+@NoArgsConstructor
+public class ImageUploadRequest {
+
```
request 자체에 대한 Schema도 작성하면 좋습니다.

juno-soodal commented on src/main/java/com/parkez/common/image/dto/ImageUploadRequest.java:
```diff
@@ -0,0 +1,32 @@
+    private Long targetId;
+
+
+    @Builder
+    public ImageUploadRequest(String targetType, Long targetId) {
+        this.targetType = targetType;
+        this.targetId = targetId;
+    }
```
빌더를 적용했으면 생성자 자체는 외부에서 직접 생성하지 못하게 private으로 막아주면 좋습니다.

juno-soodal commented on src/main/java/com/parkez/common/image/enums/AllowedExtension.java:
```diff
@@ -0,0 +1,26 @@
+    JPG("jpg"),
+    JPEG("jpeg"),
+    PNG("png");
+
+    private final String extension;
+
```
저희가 enum에서 필드값이 하나인 경우는 description으로 통일하기로 해서 수정해주세요

juno-soodal commented on src/main/java/com/parkez/common/image/S3ImageService.java:
```diff
@@ -0,0 +1,153 @@
+        return String.format("https://%s.s3.%s.amazonaws.com/%s", bucket, s3Client.getRegionName(), key);
+    }
+
+    private boolean isAllowedExtension(MultipartFile file){
+        String extension = getFileExtension(file);
+        return AllowedExtension.isAllowedExtension(extension);
+    }
```
메서드 이름이 중복해서 뭔가 어색합니다.
isAllowedExtension 메서드가 검증 하는 역할로 보이는데
확장자를 추출하고
AllowedExtension.isAllowedExtension가 검증을 한다? 이런 느낌이여서
AllowedExtension에 포함되는지로 메서드명 바꾸면 좋아보입니다.

juno-soodal commented on src/main/java/com/parkez/common/image/exception/ImageErrorCode.java:
```diff
@@ -0,0 +1,23 @@
+@RequiredArgsConstructor
+public enum ImageErrorCode implements ErrorCode {
+
+    IMAGE_NOT_FOUND("IMAGE_NOT_FOUND", HttpStatus.NOT_FOUND, "해당 폴더에 이미지가 존재하지 않습니다."),
+    IMAGE_UPLOAD_FAIL("IMAGE_UPLOAD_FAIL", HttpStatus.INTERNAL_SERVER_ERROR , "서버 오류로 인해 이미지 업로드가 실패되었습니다."),
+    INVALID_IMAGE_TYPE("INVALID_IMAGE_TYPE",HttpStatus.BAD_REQUEST , "유효하지 않는 이미지 타입입니다."),
+    IMAGE_IS_NULL("IMAGE_IS_NULL", HttpStatus.BAD_REQUEST, "이미지 파일이 비어있습니다."),
+    INVALID_EXTENSION_TYPE("INVALID_EXTENSION_TYPE", HttpStatus.BAD_REQUEST, "허용되지 않는 파일 확장자입니다.");
```
이 부분도 팀 규칙에 맞게 수정해주세요

juno-soodal commented on src/test/java/com/parkez/image/service/S3ImageServiceTest.java:
```diff
@@ -0,0 +1,198 @@
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+public class S3ImageServiceTest {
```
테스트 코드를 작 작성해주셨습니다.
다만 동일한 메서드에서 성공 케이스 , 실패 케이스가 여러가지가 있기 때문에
@Nested 클래스를 사용하여 케이스별 테스트 메서드를 하나의 클래스로 묶으면 테스트 코드 가독성이 더욱 좋아질 것 같습니다.

juno-soodal commented on src/main/java/com/parkez/common/image/S3ImageService.java:
```diff
@@ -0,0 +1,153 @@
+
+    private void uploadToS3(MultipartFile file, String fileName) {
+        try{
+            // 메타데이터 설정
+            ObjectMetadata metadata = new ObjectMetadata();
+            metadata.setContentType(file.getContentType());
+            metadata.setContentLength(file.getSize());
+
```
메서드명은 uploadToS3로 되어 있는데
이 로직은 메타데이터를 만들어서 별도의 메서드로 빼면 좋아보입니다.
그러면 주석을 쓰지 않아도 이해할 수 있습니다.
try 문에  없어도 될거 같네요

juno-soodal commented on src/main/java/com/parkez/common/image/S3ImageService.java:
```diff
@@ -0,0 +1,153 @@
+            // S3에 파일 업로드
+            s3Client.putObject(putObjectRequest);
+
+        } catch (IOException e) {
+            throw new ParkingEasyException(ImageErrorCode.IMAGE_UPLOAD_FAIL);
+        }
```
캐치 문으로 IOException e 을 잡아주셨는데
지금 예외 자체는 저희가 커스텀한 예외로 변환해서 던지고 있습니다.
커스텀 예외는 기존 예외를 포함해서 던지고있지 않기때문에 log로 어떤 예외가 발생했는지 작성해주세요.
그리고 IOException 하나로 잡는거보다는 좀 더 상세하게 s3 통신 에러인지 디테일 하게 잡아주면 좋습니다.

juno-soodal reviewed (changes requested):
수고하셨습니다.

minjonyyy committed:
refactor : 팀 규칙에 맞게 순서 변경, 에러코드 이름 변경 #43

minjonyyy committed:
add : 스웨거 @Operation 추가, refactor : 서비스 로직 인터페이스 의존으로 수정 #43

minjonyyy committed:
refactor/style : 요청 dto 패키지 분리 #43

minjonyyy committed:
refactor/style : 요청 dto 패키지 분리 #43

minjonyyy committed:
refactor : url 응답 dto 추가 #43

minjonyyy committed:
refactor : 메서드명 변경 #43

minjonyyy committed:
refactor : 얼리 리턴 형태로 변경하여 가독성 높임, 메서드명 변경 #43

minjonyyy committed:
refactor : builder 적용 및 접근제어자 수정 #43

minjonyyy committed:
refactor : 메서드명 직관적으로 변경 #43

minjonyyy committed:
feat/refactor : 예외처리 구체적으로 추가, 메타데이터 생성 메서드 분리 #43

minjonyyy committed:
Merge branch 'dev' into feat/image

minjonyyy committed:
refactor : 프로젝트 구조 변경 #43

minjonyyy committed:
refactor : 일관성 유지를 위해 @RequiredArgsContructor 추가 #43

minjonyyy committed:
refactor/fix : dto에서 String이 아닌 enum으로 요청받도록 수정, 기존 메서드 제거 #43

minjonyyy committed:
refactor : request dto 하나로 통일  #43

minjonyyy committed:
refactor : 메서드명 replace로 변경  #43

minjonyyy committed:
refactor : 피드백 반영하여 리팩토링 #43

1. request dto 하나로 통일
2. 파일명 생성 시 LocalDatetime으로 변경
3. 파일명 생성 시 string.fomat으로 변경
4. 메서드명 replace로 변경

minjonyyy committed:
test : 피드백 반영하여 테스트코드 리팩토링 #43

서비스로직 테스트 커버리지(라인) 100%

minjonyyy committed:
style : @Schema example 내용 수정, 사용하지 않는 dto 삭제 #43
